lib/base/dependencygraph.hpp

```diff
@@ -18,15 +18,25 @@ namespace icinga {
 class DependencyGraph
 {
 public:
-	static void AddDependency(Object* child, Object* parent);
-	static void RemoveDependency(Object* child, Object* parent);
-	static std::vector<Object::Ptr> GetChildren(const Object::Ptr& parent);
+	static void AddDependency(ConfigObject* child, ConfigObject* parent);
+	static void RemoveDependency(ConfigObject* child, ConfigObject* parent);
+	static std::vector<ConfigObject::Ptr> GetChildren(const ConfigObject::Ptr& parent);
+
+	static void AddDependency(ObjectImpl<ConfigObject>* child, ObjectImpl<ConfigObject>* parent)
+	{
+		AddDependency(static_cast<ConfigObject*>(child), static_cast<ConfigObject*>(parent));
+	}
+
+	static void RemoveDependency(ObjectImpl<ConfigObject>* child, ObjectImpl<ConfigObject>* parent)
+	{
+		RemoveDependency(static_cast<ConfigObject*>(child), static_cast<ConfigObject*>(parent));
+	}
```

Comment:
Why not perform this static cast directly in the class compiler instead of adding such overloads here?

Reply:
You mean to explicitly static cast every time they're called, both in mkclass-generated code and in every .ti file using the DG?

Reply:
The second argument should always be of type `ConfigObject::Ptr`, you just need to static cast `this` within the class compiler while generating the generic code.

Reply:
> You mean to explicitly static cast every time they're called, both in mkclass-generated code and in every .ti file using the DG?

No, like I said, you just need to cast the `this` argument in the dynamically generated code by mkclass and remove those overloads. I'm not sure what you're referring to with _every time they're called_, since you're basically doing the same thing with those overloads. And what does DG stand for?

Reply:
In fact the casting is only needed within `void ObjectImpl<ConfigObject>::TrackZoneName(const String& oldValue, const String& newValue)` in `confiobject-ti.cpp`, all the other usages don't even need such type casting.

Reply:
> The second argument should always be of type `ConfigObject::Ptr`, you just need to static cast `this` within the class compiler while generating the generic code.

Now my methods look a bit asymmetric, but ok.

Reply:
> In fact the casting is only needed within `void ObjectImpl<ConfigObject>::TrackZoneName(const String& oldValue, const String& newValue)` in `confiobject-ti.cpp`, all the other usages don't even need such type casting.

# 💡